submissions/iakow/OOP-exercise/index.js

```diff
@@ -0,0 +1,83 @@
+class Inhabitant {
+   constructor({ species, name, say, gender }) {
+      this.species = species;
+      this.name = name;
+      this.say = say;
+      this.gender = gender;
+      this.friends = new Set();
+   }
+
+   makeFriends(...applicants) {
+      applicants.filter(applicant => applicant instanceof Inhabitant && applicant != this)
+         .forEach(applicant => this.friends.add(applicant));
+   }
+
+   toString() {
+      const { name, species, friends, gender, say } = this;
+
+      const friendNames = [];
+      friends.forEach((friend) => friendNames.push(friend.name));
+      const friendsNamesString = friendNames.length ? `${friendNames.join(', ')};` : '';
+
+      return `${species}; ${name}; ${gender}; ${say}; ${friendsNamesString}`;
+   }
+}
+
+class Animal extends Inhabitant {
+   constructor({ species, name, gender, say, paws = 4 }) {
+      super({ species, name, say, gender });
+      this.paws = paws;
+   }
+
+   toString() {
+      return `${super.toString()} ${this.paws};`
+   }
+}
+
+class Dog extends Animal {
+   constructor({ name, gender, paws }) {
+      super({ species: 'dog', name, gender, say: 'WOOF-WOOF!!!', paws });
+   }
+}
+
+class Cat extends Animal {
+   constructor({ name, gender, paws }) {
+      super({ species: 'cat', name, gender, say: 'Meow...', paws });
+      this.friends.add(this);
+   }
+
+   bite(victim) {
+      victim.species += this.species;
+      victim.say = this.say;
+   }
+}
+
+class Human extends Inhabitant {
+   constructor({ species = "human", name, say, gender, hands = 2, legs = 2 }) {
+      super({ species, name, say, gender });
+      this.legs = legs;
+      this.hands = hands;
+   }
+
+   toString() {
+      return `${super.toString()} ${this.hands}, ${this.legs};`;
```

Comment:
You may want here to join an array of properties for code style consistency and for the sake of scalability.

Not required to fix now for you'd need to resolve inconsistency with friends list that is always terminated with a semicolon.

Reply:
I left it as it is, but the pattern took it into use, thanks.